flink/v1.16/flink/src/test/java/org/apache/iceberg/flink/source/enumerator/TestContinuousIcebergEnumerator.java

```diff
@@ -147,6 +148,85 @@ public void testRequestingReaderUnavailableWhenSplitDiscovered() throws Exceptio
+
+    // register reader-2, and let it request a split
+    enumeratorContext.registerReader(2, "localhost");
+    enumerator.addReader(2);
+    enumerator.handleSourceEvent(2, new SplitRequestEvent());
+
+    // add splits[0] to the planner for next discovery
+    splitPlanner.addSplits(Arrays.asList(splits.get(0)));
+    enumeratorContext.triggerAllActions();
+
+    // because discovered split was assigned to reader, pending splits should be empty
+    Assert.assertEquals(0, enumerator.snapshotState(1).pendingSplits().size());
+    // split assignment to reader-2 should contain splits[0, 1)
+    Assert.assertEquals(
+        splits.subList(0, 1), enumeratorContext.getSplitAssignments().get(2).getAssignedSplits());
+
+    // add the remaining 9 splits (one for every snapshot)
+    // run discovery cycles while reader-2 still processing the splits[0]
+    for (int i = 1; i < 10; ++i) {
+      splitPlanner.addSplits(Arrays.asList(splits.get(i)));
+      enumeratorContext.triggerAllActions();
+    }
+
+    // can only discover up to 3 snapshots/splits
+    Assert.assertEquals(3, enumerator.snapshotState(2).pendingSplits().size());
+    // split assignment to reader-2 should be splits[0, 1)
+    Assert.assertEquals(
+        splits.subList(0, 1), enumeratorContext.getSplitAssignments().get(2).getAssignedSplits());
+
+    // now reader-2 finished splits[0]
+    enumerator.handleSourceEvent(2, new SplitRequestEvent(Arrays.asList(splits.get(0).splitId())));
+    enumeratorContext.triggerAllActions();
+    // still have 3 pending splits. After assigned splits[1] to reader-2, one more split was
+    // discovered and added.
+    Assert.assertEquals(3, enumerator.snapshotState(3).pendingSplits().size());
+    // split assignment to reader-2 should be splits[0, 2)
+    Assert.assertEquals(
+        splits.subList(0, 2), enumeratorContext.getSplitAssignments().get(2).getAssignedSplits());
+
+    // run 3 more split discovery cycles
+    for (int i = 0; i < 3; ++i) {
+      enumeratorContext.triggerAllActions();
+    }
+
+    // no more splits are discovered due to throttling
+    Assert.assertEquals(3, enumerator.snapshotState(4).pendingSplits().size());
+    // split assignment to reader-2 should still be splits[0, 2)
+    Assert.assertEquals(
+        splits.subList(0, 2), enumeratorContext.getSplitAssignments().get(2).getAssignedSplits());
+
+    // now reader-2 finished splits[1]
+    enumerator.handleSourceEvent(2, new SplitRequestEvent(Arrays.asList(splits.get(1).splitId())));
+    enumeratorContext.triggerAllActions();
+    // still have 3 pending splits. After assigned new splits[2] to reader-2, one more split was
+    // discovered and added.
+    Assert.assertEquals(3, enumerator.snapshotState(5).pendingSplits().size());
+    // split assignment to reader-2 should be splits[0, 3)
+    Assert.assertEquals(
+        splits.subList(0, 3), enumeratorContext.getSplitAssignments().get(2).getAssignedSplits());
```

Comment:
Seems like a duplicated check - maybe use a method for this?

Reply:
I think you are talking about the duplication btw 199-206 and 220-227. We can extract a method to save the some code duplication. I didn't do it for the following reasons.
1. it is 5 lines of code
2. the readability may not be as easy when calling a method like `runOneCycle(int subtask, int pendingSplitCount, List<Integer> expectedSplitAssignment, long checkpointId)`, as it is hard to see what the integer arg means.

But I am not very opinionated here. Please let me know if you think the method extraction is better here.

Reply:
Same here - no strong feelings about it on a second glance -, but my first impression was that there seems to me plenty of occurrences of this block, just with different parameters:
```
    Assert.assertEquals(3, enumerator.snapshotState(3).pendingSplits().size());
    Assert.assertEquals(
        splits.subList(0, 2), enumeratorContext.getSplitAssignments().get(2).getAssignedSplits());
```

Reply:
I will keep it as it is then for better readability